cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillLauncher.java

```diff
@@ -157,4 +166,70 @@ private void deletePod(PodInfo podInfo) {
       LOG.warn("Failed to delete pod {} with uid {}", podInfo.getName(), podInfo.getUid(), e);
     }
   }
+
+  /**
+   * Get instance id from the annotation. The annotation will have array of boolean
+   */
+  private InstanceInfo getInstanceInfo(PodInfo podInfo) throws Exception {
+    int instanceId = 0;
+    int instanceCount = 1;
+
+    if (podInfo.getOwnerReferences().stream().anyMatch(r -> "Job".equals(r.getKind()))) {
+      String jobName = podInfo.getOwnerReferences().get(0).getName();
+      ApiClient client = Config.defaultClient();
+      BatchV1Api batchApi = new BatchV1Api(client);
+      Random rand = new Random();
+
+      try {
+        boolean done = false;
+        int randomNumber = 0;
+
+        while (!done) {
+          V1Job v1Job = batchApi.readNamespacedJob(jobName, podInfo.getNamespace(), null, null, null);
+          Map<String, String> annotations = v1Job.getMetadata().getAnnotations();
+          if (!annotations.containsKey(JOB_INSTANCES)) {
+            // this means parallelism is 1. No further processing is needed
+            done = true;
+            continue;
+          }
+          boolean[] jobInstances = GSON.fromJson(annotations.get(JOB_INSTANCES), boolean[].class);
+          do {
+            // make sure we choose index that is still unreserved yet
+            randomNumber = rand.nextInt(jobInstances.length);
+          } while ((jobInstances[randomNumber]));
+
+          // reserve the chosen index
+          jobInstances[randomNumber] = true;
+          annotations.put(JOB_INSTANCES, GSON.toJson(jobInstances));
+          v1Job.getMetadata().setAnnotations(annotations);
+
+          try {
+            batchApi.replaceNamespacedJob(jobName, podInfo.getNamespace(), v1Job, null, null, null);
+            instanceCount = jobInstances.length;
+            done = true;
+          } catch (ApiException e) {
+            if (e.getCode() == 409) {
```

Comment:
Would be good to add a comment on when this happens.

For correctness I assume it happens if the annotations changed after you read the job and before the replace() call happens, but I don't see how k8s would know that. Is there some revision number somewhere or something like that?

Reply:
Yes, replace apis has a V1Job parameter. With that, k8s can do resource version check since each resource will have specific resource version.  Here is an article I referred to about conflict detection in k8s:
https://hinyinlam.medium.com/k8s-java-client-sdk-patching-and-update-a-resource-3b42de7f69ed

However, patch api does not support optimistic locks. Thats why currently we are using replace apis.